### PR TITLE
Top-sorting code

### PR DESCRIPTION
**Pull request by danpovey**
this hasn't been compiled.  I was hoping someone else might be able to take over from here.

### CODE REVIEW

**Comment by danpovey** [comment]
OK, this is a rewritten version of TopSort which should be correct from a very high-level point of view but probably has tons of bugs.  I was hoping someone else might be able to help compile and test it?  Anyway it's good to have others familiar with the coding style and the kinds of methods I'm using.

**Comment by qindazhu** [comment]
ok，i will do this tomorrow as the start point of debugging fsa related algorithms.

File: k2/csrc/fsa_utils.cu [hunk 1]
```diff
@@ -515,7 +515,10 @@ Fsa FsaFromString(const std::string &s, bool openfst /*= false*/,
 std::string FsaToString(const Fsa &fsa, bool openfst /*= false*/,
                         const Array1<int32_t> *aux_labels /*= nullptr*/) {
   K2_CHECK_EQ(fsa.NumAxes(), 2);
-  K2_CHECK_EQ(fsa.Context()->GetDeviceType(), kCpu);
+  if (fsa.Context()->GetDeviceType() != kCpu) {
+    Fsa cpu_fsa = fsa.To(GetCpuContext());
+    return FsaToString(cpu_fsa, openfst, aux_labels);
```
Comment (csukuangfj) [inline k2/csrc/fsa_utils.cu:520]: aux_labels is on GPU if not NULL.

**Comment by danpovey** [comment]
FYI, I added a bunch more draft code here, of GetBatches().  Haven't looked over it carefully or tried to compile yet.

**Comment by qindazhu** [comment]
@danpovey , if you think it's ready to merge, you can merge it first, then I'll  fix bugs and test based on your merged code (so that we can avoid future conflicts)

**Comment by danpovey** [comment]
@qindazhu this doesn't compile... I'm making it so it compiles at least, then I'll merge.

**Comment by danpovey** [comment]
This is quite problematic but we'll diverge too much if I dont merge, so will merge now

**Comment by csukuangfj** [comment]
I assumed that you have not started since you just implemented the `Renumbering`.

**Comment by qindazhu** [comment]
Yes, I have not started. Will debug `GetDestStates` and `GetBatches` today.

**Pull request merged**